### PR TITLE
Fix issues found when deploying dogfood

### PR DESCRIPTION
**Pull request by ReaNAiveD**

### CODE REVIEW

File: API/common/auth.py [hunk 1]
```diff
@@ -66,24 +66,17 @@ def wrapper(*args, **kwargs) -> HttpResponse:
...
-
-    app = msal.ConfidentialClientApplication(client_id, authority, client_credential)
+def _get_auth_token(*scopes: str):
     try:
-        result = app.acquire_token_for_client(scopes=scope)
-        return result['access_token']
+        credential = DefaultAzureCredential()
+        token = credential.get_token(*scopes)
+        token = "Bearer " + token.token
+        return token
```
Comment (zhoxing-ms) [inline API/common/auth.py:74]: May I ask why do we need to modify the way of obtaining auth token?

File: API/CopilotService/__init__.py [hunk 2]
```diff
@@ -151,16 +161,28 @@ async def _build_task_context(raw_task, token):
...
     context.custom_context.estimated_task_list_tokens = num_tokens_from_message(task_list)
     context.custom_context.estimated_usage_context_tokens = num_tokens_from_message(usage_context)
     if task_list:
         guiding_steps_separation = "\nHere are the steps you can refer to for this question:\n"
-        question = question + guiding_steps_separation + str(task_list) + '\n'
+        question = _try_add_steps_to_queston(question, guiding_steps_separation, task_list, token_limit)
     if usage_context:
         commands_info_separation = "\nBelow are some potentially relevant CLI commands information as context, please select the commands information that may be used in the scenario of the question from context, and supplement the missing commands information of context\n"
-        question = question + commands_info_separation + str(usage_context)
+        question = _try_add_steps_to_queston(question, commands_info_separation, usage_context, token_limit)
```
Comment (zhoxing-ms) [inline API/CopilotService/__init__.py:177]: Have we discarded the chunks with the lowest correlation score from knowledge index when the token exceeds the limit?

File: API/common/auth.py (hunk 1 above)
Comment (ReaNAiveD) [reply API/common/auth.py:74]: The previous code did not generate a valid token. This method of authentication is much simpler.

File: API/CopilotService/__init__.py (hunk 2 above)
Comment (ReaNAiveD) [reply API/CopilotService/__init__.py:177]: The retrieved chunks are sorted by similarity scores by `learn knowledge`. The service keeps the sorted order, but filters out some irrelevant chunks. Also, we haven't found a way to resort the chunks efficiently and accurately.

Comment (zhoxing-ms) [reply API/CopilotService/__init__.py:177]: Yes, I am considering whether this is the time to introduce the fine sorting.

File: API/CopilotService/__init__.py [hunk 3]
```diff
@@ -56,15 +57,24 @@ def main(req: func.HttpRequest, context: func.Context) -> func.HttpResponse:
...
             gpt_result = gpt_generate(context, system_msg, question, history)
             result = [_build_scenario_response(gpt_result)] if gpt_result else []
 
         elif service_type == ServiceType.MIX:
-            result = knowledge_search(question, top_num)
+            try:
+                result = knowledge_search(question, top_num)
+            except HttpResponseError as e:
+                logger.error('Error from knowledge search: \n%s', e)
+                result = []
```
Comment (zhoxing-ms) [inline API/CopilotService/__init__.py]: I believe the same logic is also needed when `service_type == ServiceType.KNOWLEDGE_SEARCH`,

Comment (ReaNAiveD) [reply API/CopilotService/__init__.py]: That makes sense. Maybe we should return a status code `500` when `service_type == ServiceType.KNOWLEDGE_SEARCH` in this case.

Comment (zhoxing-ms) [reply API/CopilotService/__init__.py]: Agree

**Review by zhoxing-ms — APPROVED** [review]

File: API/CopilotService/__init__.py [hunk 4]
```diff
@@ -48,23 +49,36 @@ def main(req: func.HttpRequest, context: func.Context) -> func.HttpResponse:
...
         if service_type == ServiceType.KNOWLEDGE_SEARCH:
-            result = knowledge_search(question, top_num)
-            if len(result) == 0 or not pass_verification(context, question, result):
-                result = []
+            try:
+                result = knowledge_search(question, top_num)
+                if len(result) == 0 or not pass_verification(context, question, result):
+                    result = []
+            except HttpResponseError as e:
+                logger.error('Response Status 500: Error from knowledge search: \n%s', e, e)
```
Comment (zhoxing-ms) [inline API/CopilotService/__init__.py]: >  e, e)

Why do we need to repeat the `e`

Comment (ReaNAiveD) [reply API/CopilotService/__init__.py]: The second `e` arg is for exc_info, which could help log the trace of the error.

Comment (ReaNAiveD) [reply API/CopilotService/__init__.py]: Fix the error with `exc_info=e`

**Pull request merged**